### PR TITLE
issue #974 Implementation of Drag & Drop for files/folders within Theia

### PR DESCRIPTION
**Pull request by lmcbout**
Signed-off-by: Jacques Bouthillier <jacques.bouthillier@ericsson.com>

### CODE REVIEW

**Comment by lmcbout** [comment]
Drag & Drop files within Theia ready to be reviewed.

**Comment by akosyakov** [comment]
@lmcbout please review https://github.com/lmcbout/theia/pull/2 and merge

**Comment by lmcbout** [comment]
@akosyakov
Code rebased and merged.
Removing the "this.refresh" after the move , the Theia navigator does not get refresh automatically, you need to click to close + re-open the folder to see the move modification. previous comment said about that: " get rid of calling refresh after the move by fixing computation of affected file tree nodes on filesystem changes, we use unidirectional flow pattern, ". It has to be fixed prior to commit this PR, otherwise, there is a problem.

**Comment by lmcbout** [comment]
To explain my previous comment, not having the refresh , we see the file or folders move does not show until we force a refresh by closing and re-opening the folder.
In the following example, we see a file (EgerritColor.png ) moving from one folder (Antoine) to folder (ouf)
Second refresh missing, take the same file and move it back to the original folder
Third test, move a folder (Antoine) to a root folder (yyy).
draganddropnotrefeshed

**Comment by akosyakov** [comment]
Have you checked it with my changes? I thought that I had fixed it. I will try later locally to reproduce again.

**Comment by lmcbout** [comment]
I rebased all changes  this morning and tried again, same result, check the attachment above

**Comment by marcdumais-work** [comment]
I just tried the tip of the branch "lmcbout:issue_974", including @akosyakov 's commit on top (458718e6438dc9b7efe015056866b8e1a90a378b). I can confirm that, on my Ubuntu machine,  I see the same refresh issue as in the video above, a majority (but not all) of the time I when move a folder within the navigator, by dragging it.

**Comment by akosyakov** [comment]
I cannot reproduce it on Mac, it can be related to events generated by nfsw on Linux, do you use Linux right?

**Comment by akosyakov** [comment]
Found an issue for it: https://github.com/Axosoft/nsfw/issues/42. As the workaround for a move within Theia, we can call refresh on a parent of a target node, but generally, it won't work on Linux till the issue if fixed.

File: packages/filesystem/src/browser/file-tree/file-tree-model.ts [hunk 1]
```diff
@@ -111,6 +110,17 @@ export class FileTreeModel extends TreeModel implements LocationService {
...
     }
 
+    /**
+     * Move the given source file or directory to the given target directory.
+     */
+    async move(source: ITreeNode, target: ITreeNode) {
+        if (DirNode.is(target) && FileStatNode.is(source)) {
+            const sourceUri = source.uri.toString();
+            const targetUri = target.uri.resolve(source.name).toString();
+            await this.fileSystem.move(sourceUri, targetUri, { overwrite: true });
```
Comment (akosyakov) [inline packages/filesystem/src/browser/file-tree/file-tree-model.ts]: to workaround https://github.com/theia-ide/theia/pull/975#issuecomment-352356821 hereafter `move`:
```ts
// to workaround https://github.com/Axosoft/nsfw/issues/42
this.refresh(target);
```

**Comment by lmcbout** [comment]
just rebase and add the refresh comment

**Comment by akosyakov** [comment]
@lmcbout LGTM

**Comment by lmcbout** [comment]
Appveyor known error about git, so I am pushing the code.

**Pull request merged**